### DISCUSSION TITLE
Automatically authorize OAuth Client on running a VM within a system already authorized

### DISCUSSION THREAD

**Discussion by walt-jones** [keycloak/keycloak#35721] [Q&A]
We need a way to authorize an OAuth Client running on a VM within the system that did the initial authorization.  Making a Client public access type allows us to avoid having to input username/password again for OAuth on the same system, but when we're inside a VM we can't seem to find a way to allow an OAuth Client access without being prompted again for the same username/password already input on the parent system.  We started looking into things like https://www.keycloak.org/securing-apps/token-exchange but there's no obvious way to taken an existing auth and programmatically give out a token to a Client running on another system. 

We understand conceptually why the VM is prompting for credentials, but need to find a way to reuse an authorization so people aren't signing in multiple times.

How can we make this work?

**Comment by darius-m** [comment]
Could you please elaborate on how this would work (or give an example for a similar system)? Theoretically a VM is a separate system altogether, so I'm not sure how this would work - do you have an API from the VM to its parent / other system (e.g., similar to how cloud-init works) to fetch the token from one system to another? In theory you should be able to just pass (reuse) the bearer token around multiple clients / instances, if the audience value is correct, but I'm not sure how your flow should look like.

**Comment by walt-jones** [comment]
Yes, we have an API and other channels between the VM and the parent, so can move anything securely between the two, including bearer tokens and other data.  How would reusing the bearer token between Clients work?

**Reply by embesozzi** [reply]
We should split the process into two stages: `authentication` and `authorization`.

Regarding the `authentication` stage, to avoid requesting credentials (login) between multiple clients, this is called Single Sign-On (SSO). However, if you follow the OpenID Connect standard, the only way to achieve this is through browser-based authentication, where the IDP cookie is still valid.

In the `authorization` space, let’s say your APIs are protected by OAuth 2.0, acting as a resource server. In this case, you can simplify the architecture by using a broader audience or multiple audiences in token, allowing other clients to accept the token. Another approach is token exchange, where one use case is creating a new token for  a new client.  Here is a [disscusion](https://github.com/keycloak/keycloak/discussions/26502) about the different use cases in token exchange by Thomas.

**Comment by walt-jones** [comment]
Thanks @embesozzi. We've been looking at token exchange as a way to handle this but don't know enough about the inner workings of OAuth to implement a basic test to confirm this as a viable option.  Can you advise on a basic setup for grabbing a usable token from an existing auth through KC API and handing it to another Client that's using OpenID Connect... or perhaps point us to something that's been done similar enough to this that we can piece together a basic process?  We can automate what's needed through our own APIs once we've got the basic idea working.
